mantis/solver/src/solver.rs

```diff
@@ -0,0 +1,83 @@
+
+pub struct Solver {
+    orders: OrderList,
+    target_price: f64,
+    buy_token: f64,
+    sell_token: f64,
+    order: Order,
+}
+
+impl Solver {
+    pub fn new(orders: OrderList, target_price: f64, buy_token: f64, sell_token: f64) -> Self {
+        Self {
+            orders,
+            target_price,
+            buy_token,
+            sell_token,
+            order: Order::new(0.0, 0.0, OrderType::Buy, "fake-order".to_string()),
+        }
+    }
+
+    pub fn limit_price(&self) -> f64 {
+        self.target_price
+    }
+
+    fn f_maximize(&self, order: &Order) -> f64 {
+        match order.order_type {
+            OrderType::Buy => {
+                self.buy_token
+                    - order.amount_filled
+                    + (self.sell_token + order.amount_out) * self.target_price
+            }
+            OrderType::Sell => {
+                (self.buy_token + order.amount_out) / self.target_price + self.sell_token - order.amount_filled
```

Comment:
Why are we multiplying by target price in one branch but dividing in the other?

In fact the expressions don’t make sense looking at the units.  In Buy branch we have `amount + amount * price` and in the Sell branch we have `amount / price + amount`.

Reply:
```python
def f_maximaize(self, order: Order):
        if order.type is OrderType.BUY:
            return (
                self.buy_token
                - order.amount_filled
                + (self.sell_token + order.amount_out) * self.target_price
            )
        return (
            (self.buy_token + order.amount_out) / self.target_price
            + self.sell_token
            - order.amount_filled
        )
```

this is direct port from Bruno codebase. Are you saying it is wrong?

Reply:
I don’t know if it’s wrong but it certainly feels wrong to me. Perhaps need to ask Bruno about it.

Reply:
origi nal version of  code was correct in that area as i recall. than there was refactroing. i pprted code as blackbox. will find out if it is weird on contract simulation